libs/core-ui/src/lib/util/StatisticsUtils.ts

```diff
@@ -144,18 +147,107 @@ const generateRegressionStats: (
   ];
 };
 
-const generateMulticlassStats: (outcomes: number[]) => ILabeledStatistic[] = (
+const generateMulticlassStats: (
+  trueYs: number[],
+  predYs: number[],
+  outcomes: number[]
+) => ILabeledStatistic[] = (
+  trueYs: number[],
+  predYs: number[],
   outcomes: number[]
 ): ILabeledStatistic[] => {
   const correctCount = outcomes.filter(
     (x) => x === MulticlassClassificationEnum.Correct
   ).length;
   const total = outcomes.length;
+
+  const predLabels = new Set();
+  predYs.forEach((predY: number) => {
+    if (!predLabels.has(predY)) {
+      predLabels.add(predY);
+    }
+  });
+  const precisionScores: number[] = [];
+  predLabels.forEach((label) => {
+    let truePositives = 0;
+    let totalPositives = 0;
+    predYs.forEach((predY, index) => {
+      if (predY === label) {
+        totalPositives += 1;
+        if (trueYs[index] === label) {
+          truePositives += 1;
+        }
+      }
+    });
+    const precision = truePositives / totalPositives;
+    precisionScores.push(precision);
+  });
+
+  const macroPrecision =
+    precisionScores.reduce((total, curr) => {
+      return total + curr;
+    }, 0) / precisionScores.length;
+
+  const trueLabels = new Set();
+  trueYs.forEach((trueY: number) => {
+    if (!trueLabels.has(trueY)) {
+      trueLabels.add(trueY);
+    }
+  });
+  const recallScores: number[] = [];
+  trueLabels.forEach((label) => {
+    let truePositives = 0;
+    let totalTruths = 0;
+    trueYs.forEach((trueY, index) => {
+      if (trueY === label) {
+        totalTruths += 1;
+        if (predYs[index] === label) {
+          truePositives += 1;
+        }
+      }
+    });
+    const recall = truePositives / totalTruths;
+    recallScores.push(recall);
+  });
+
+  const macroRecall =
+    recallScores.reduce((total, curr) => {
+      return total + curr;
+    }, 0) / recallScores.length;
+
+  const f1Scores: number[] = [];
+
+  precisionScores.forEach((precision, index) => {
+    const recall = recallScores[index];
+    const f1 = 2 * ((precision * recall) / (precision + recall)) || 0;
```

Comment:
I feel like there is a lot of code in other places, like StatsAggregator.ts, that does this already.  I wonder if we can refactor out some of these statistics computations.